File: adapter/typescript/Adapter.ts
// Target interface
interface MediaPlayer {
  play(filename: string): string;
}

// Adaptee interface
interface AdvancedMediaPlayer {
  playVLC(filename: string): string;
  playMP4(filename: string): string;
}

// Concrete Adaptee
class VLCPlayer implements AdvancedMediaPlayer {
  playVLC(filename: string): string {
    return `Playing VLC file: ${filename}`;
  }

  playMP4(filename: string): string {
    return '';
  }
}

class MP4Player implements AdvancedMediaPlayer {
  playVLC(filename: string): string {
    return '';
  }

  playMP4(filename: string): string {
    return `Playing MP4 file: ${filename}`;
  }
}

// Adapter
class MediaAdapter implements MediaPlayer {
  private advancedPlayer: AdvancedMediaPlayer;

  constructor(audioType: string) {
    if (audioType === 'vlc') {
      this.advancedPlayer = new VLCPlayer();
    } else if (audioType === 'mp4') {
      this.advancedPlayer = new MP4Player();
    }
  }

  play(filename: string): string {
    if (this.advancedPlayer instanceof VLCPlayer) {
      return this.advancedPlayer.playVLC(filename);
    } else if (this.advancedPlayer instanceof MP4Player) {
      return this.advancedPlayer.playMP4(filename);
    }
    return '';
  }
}

// Concrete MediaPlayer using adapter
class AudioPlayer {
  play(audioType: string, filename: string): string {
    if (audioType === 'mp3') {
      return `Playing MP3 file: ${filename}`;
    } else if (audioType === 'vlc' || audioType === 'mp4') {
      const adapter = new MediaAdapter(audioType);
      return adapter.play(filename);
    }
    return `Invalid media type: ${audioType}`;
  }
}

// Usage
const player = new AudioPlayer();
console.log(player.play('mp3', 'song.mp3'));
console.log(player.play('vlc', 'movie.vlc'));
console.log(player.play('mp4', 'video.mp4'));

File: adapter/java/Adapter.java
// Target interface
interface MediaPlayer {
    String play(String filename);
}

// Adaptee (old interface)
interface AdvancedMediaPlayer {
    String playVLC(String filename);
    String playMP4(String filename);
}

// Concrete Adaptee: VLC Player
class VLCPlayer implements AdvancedMediaPlayer {
    @Override
    public String playVLC(String filename) {
        return "Playing VLC file: " + filename;
    }

    @Override
    public String playMP4(String filename) {
        return "";
    }
}

// Concrete Adaptee: MP4 Player
class MP4Player implements AdvancedMediaPlayer {
    @Override
    public String playVLC(String filename) {
        return "";
    }

    @Override
    public String playMP4(String filename) {
        return "Playing MP4 file: " + filename;
    }
}

// Adapter
class MediaAdapter implements MediaPlayer {
    private AdvancedMediaPlayer advancedPlayer;

    public MediaAdapter(String audioType) {
        if (audioType.equalsIgnoreCase("vlc")) {
            advancedPlayer = new VLCPlayer();
        } else if (audioType.equalsIgnoreCase("mp4")) {
            advancedPlayer = new MP4Player();
        }
    }

    @Override
    public String play(String filename) {
        if (advancedPlayer instanceof VLCPlayer) {
            return advancedPlayer.playVLC(filename);
        } else if (advancedPlayer instanceof MP4Player) {
            return advancedPlayer.playMP4(filename);
        }
        return "";
    }
}

// Concrete implementation using adapter
class AudioPlayer implements MediaPlayer {
    @Override
    public String play(String filename) {
        return "Playing MP3 file: " + filename;
    }

    public String play(String audioType, String filename) {
        if (audioType.equalsIgnoreCase("mp3")) {
            return "Playing MP3 file: " + filename;
        } else if (audioType.equalsIgnoreCase("vlc") || audioType.equalsIgnoreCase("mp4")) {
            MediaAdapter adapter = new MediaAdapter(audioType);
            return adapter.play(filename);
        }
        return "Invalid media type: " + audioType;
    }
}

// Main class demonstrating Adapter pattern
public class Adapter {
    public static void main(String[] args) {
        AudioPlayer player = new AudioPlayer();
        System.out.println(player.play("mp3", "song.mp3"));
        System.out.println(player.play("vlc", "movie.vlc"));
        System.out.println(player.play("mp4", "video.mp4"));
    }
}
